File: sdks/python/src/archon/agent.py
"""Agent — the core Archon API with full production harness."""

from __future__ import annotations

import json
import time
import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

import litellm

from archon.budget import Budget, BudgetExceeded
from archon.cache import SemanticCache
from archon.router import Router, RoutingDecision
from archon.sanitize import Sanitizer
from archon.security import PolicyAction, SecurityConfig
from archon.tool import ToolDef
from archon.trace import TraceStore
from archon.types import AgentResult, Step, Tier

_DEFAULT_TRACE_DB = Path.home() / ".archon" / "traces.db"


class Agent:
    """An AI agent with built-in cost control, routing, security, and tracing.

    Every LLM call passes through 5 gates:
    1. Policy check — is this tool call allowed?
    2. Model routing — cheapest sufficient model
    3. Execute — LLM call + sandboxed tool execution
    4. Validate — sanitize tool outputs
    5. Trace — log everything
    """

    def __init__(
        self,
        *,
        name: str,
        instructions: str,
        tools: list[ToolDef] | None = None,
        model: str = "auto",
        budget: Budget | None = None,
        max_steps: int = 25,
        security: SecurityConfig | None = None,
        cache: SemanticCache | None = None,
        trace_store: TraceStore | None = None,
        sanitize: bool = True,
    ) -> None:
        self.name = name
        self.instructions = instructions
        self.tools = tools or []
        self.model = model
        self.budget = budget or Budget()
        self.max_steps = max_steps
        self.security = security or SecurityConfig.permissive()
        self.cache = cache
        self.trace_store = trace_store
        self._sanitizer = Sanitizer(strict=sanitize) if sanitize else None
        self._router = Router()

    async def run(self, prompt: str) -> AgentResult:
        """Execute the agent loop with full production harness."""
        run_id = str(uuid.uuid4())
        result = AgentResult(run_id=run_id, output="", started_at=datetime.now(timezone.utc))

        # Register run in trace store
        if self.trace_store:
            self.trace_store.start_run(run_id, self.name, prompt)
            self.trace_store.audit(run_id, self.name, "run_started", prompt[:200])

        # Gate 0: Check semantic cache
        if self.cache:
            routing = self._resolve_model(prompt)
            cached = self.cache.get(prompt, routing.model)
            if cached:
                result.output = cached.response
                result.finished_at = datetime.now(timezone.utc)
                step = Step(
                    id=f"{run_id}-cache", model=routing.model, tier=routing.tier,
                    cost_usd=0.0, cached=True,
                )
                result.steps.append(step)
                if self.trace_store:
                    self.trace_store.record_step(self.name, run_id, step)
                    self.trace_store.audit(run_id, self.name, "cache_hit")
                    self.trace_store.finish_run(run_id, result.output, 0.0, 1, 0, 0)
                return result

        messages: list[dict[str, Any]] = [
            {"role": "system", "content": self.instructions},
            {"role": "user", "content": prompt},
        ]
        tool_schemas = [t.to_schema() for t in self.tools]
        tool_map = {t.name: t for t in self.tools}

        for step_num in range(self.max_steps):
            # Gate 1: Route to optimal model
            last_content = messages[-1].get("content", "") if messages else prompt
            routing = self._resolve_model(last_content if step_num > 0 else prompt)

            # Gate 2: Check budget
            try:
                self.budget.check(0.01)
            except BudgetExceeded:
                result.output = f"[Budget exceeded after {step_num} steps, spent ${self.budget.spent:.4f}]"
                if self.trace_store:
                    self.trace_store.audit(run_id, self.name, "budget_exceeded", f"${self.budget.spent:.4f}")
                break

            # Gate 3: Call LLM
            t0 = time.monotonic()
            response = await litellm.acompletion(
                model=routing.model,
                messages=messages,
                tools=tool_schemas if tool_schemas else None,
            )
            latency_ms = int((time.monotonic() - t0) * 1000)

            # Track cost
            usage = response.usage
            input_tokens = usage.prompt_tokens if usage else 0
            output_tokens = usage.completion_tokens if usage else 0
            cost = 0.0
            if hasattr(response, "_hidden_params"):
                cost = response._hidden_params.get("response_cost", 0.0)
            if cost == 0.0:
                try:
                    cost = litellm.completion_cost(completion_response=response)
                except Exception:
                    cost = 0.0
            self.budget.record(cost)

            step = Step(
                id=f"{run_id}-{step_num}", model=routing.model, tier=routing.tier,
                input_tokens=input_tokens, output_tokens=output_tokens,
                cost_usd=cost, latency_ms=latency_ms,
            )

            choice = response.choices[0]
            msg = choice.message

            if msg.tool_calls:
                messages.append(msg.model_dump())
                for tc in msg.tool_calls:
                    fn_name = tc.function.name
                    step.tool_call = fn_name

                    # Gate 1 (per tool): Policy check
                    args = json.loads(tc.function.arguments)
                    policy_result = self.security.policy.evaluate(fn_name, args)

                    if policy_result == PolicyAction.DENY:
                        tool_output = f"[BLOCKED by policy: {fn_name}]"
                        if self.trace_store:
                            self.trace_store.audit(run_id, self.name, "tool_denied", fn_name)
                    elif policy_result == PolicyAction.REQUIRE_APPROVAL:
                        tool_output = f"[REQUIRES APPROVAL: {fn_name}]"
                        if self.trace_store:
                            self.trace_store.audit(run_id, self.name, "tool_approval_required", fn_name)
                    elif fn_name in tool_map:
                        # Execute tool
                        try:
                            raw_result = tool_map[fn_name].fn(**args)
                            tool_output = str(raw_result)
                        except Exception as e:
                            tool_output = f"[Tool error: {e}]"
                            if self.trace_store:
                                self.trace_store.audit(run_id, self.name, "tool_error", f"{fn_name}: {e}")
                    else:
                        tool_output = f"[Unknown tool: {fn_name}]"

                    # Gate 4: Sanitize tool output
                    if self._sanitizer:
                        san = self._sanitizer.sanitize(tool_output)
                        if san.was_modified and self.trace_store:
                            self.trace_store.audit(
                                run_id, self.name, "output_sanitized",
                                f"{fn_name}: threats={san.threats_found} level={san.threat_level.value}",
                            )
                        tool_output = san.sanitized

                    messages.append({"role": "tool", "tool_call_id": tc.id, "content": tool_output})

                # Gate 5: Record step
                self._record_step(result, step, run_id)
                continue

            # Final response — no tool calls
            self._record_step(result, step, run_id)
            result.output = msg.content or ""

            # Cache the result
            if self.cache and result.output:
                self.cache.put(prompt, routing.model, result.output, result.total_cost_usd)

            break

        result.finished_at = datetime.now(timezone.utc)
        result.model_usage = {}
        for s in result.steps:
            result.model_usage[s.model] = result.model_usage.get(s.model, 0) + 1

        if self.trace_store:
            self.trace_store.finish_run(
                run_id, result.output, result.total_cost_usd,
                len(result.steps), result.total_input_tokens + result.total_output_tokens,
                result.total_latency_ms,
            )

        return result

    def _record_step(self, result: AgentResult, step: Step, run_id: str) -> None:
        result.steps.append(step)
        result.total_cost_usd += step.cost_usd
        result.total_input_tokens += step.input_tokens
        result.total_output_tokens += step.output_tokens
        result.total_latency_ms += step.latency_ms
        if self.trace_store:
            self.trace_store.record_step(self.name, run_id, step)

    def _resolve_model(self, text: str) -> RoutingDecision:
        if self.model != "auto":
            return RoutingDecision(tier=Tier.STANDARD, model=self.model, reason="explicit model")
        return self._router.route(text, self.budget.remaining)

File: sdks/typescript/src/agent.ts
/**
 * Agent — the core Archon TypeScript API.
 *
 * Every LLM call passes through 5 gates:
 *   1. Cache check — skip if we've answered this before
 *   2. Model routing — cheapest sufficient model
 *   3. Budget check — hard stop before overspending
 *   4. Execute — LLM call via OpenAI-compatible API
 *   5. Track — record cost, tokens, latency
 *
 * @example
 * ```ts
 * const agent = new Agent({
 *   name: "researcher",
 *   instructions: "Find accurate information.",
 *   model: "auto",
 *   budget: { maxPerRun: 0.50 },
 *   apiKey: process.env.OPENAI_API_KEY,
 * });
 * const result = await agent.run("What is quantum computing?");
 * console.log(result.output, result.totalCostUsd);
 * ```
 */

import { Budget, BudgetExceeded } from "./budget.js";
import { Router } from "./router.js";
import { TelemetryClient } from "./telemetry.js";
import type { AgentResult, BudgetConfig, RoutingDecision, Step, Tier, ToolDef } from "./types.js";

// ── Configuration ─────────────────────────────────────

/** Conservative cost estimate for budget pre-check (USD). */
const ESTIMATED_COST_PER_CALL = 0.01;

/** Default max steps before the agent stops. */
const DEFAULT_MAX_STEPS = 25;

/** Default OpenAI-compatible API base URL. */
const DEFAULT_API_BASE = "https://api.openai.com/v1";

// ── Agent config ──────────────────────────────────────

export interface AgentConfig {
  /** Human-readable agent name (used in traces). */
  name: string;
  /** System prompt — the agent's personality and rules. */
  instructions: string;
  /** Tools the agent can invoke. */
  tools?: ToolDef[];
  /** Model name, or "auto" for automatic routing. */
  model?: string;
  /** Budget with hard spending limits. */
  budget?: BudgetConfig;
  /** Maximum LLM calls before the agent stops. */
  maxSteps?: number;
  /** API key for the LLM provider. */
  apiKey?: string;
  /** OpenAI-compatible API base URL. */
  apiBase?: string;
  /**
   * Cloud telemetry client. Defaults to a new `TelemetryClient` that
   * auto-enables when `ARCHON_API_KEY` is set in the environment. Pass
   * a configured instance to customize the dashboard URL, or `false`
   * to explicitly disable cloud uploads even if the env var is set.
   */
  telemetry?: TelemetryClient | boolean;
}

// ── Agent class ───────────────────────────────────────

export class Agent {
  readonly name: string;
  readonly instructions: string;
  readonly tools: ToolDef[];
  readonly model: string;
  readonly budget: Budget;
  readonly maxSteps: number;
  private readonly router: Router;
  private readonly apiKey: string;
  private readonly apiBase: string;
  private readonly telemetry: TelemetryClient | undefined;

  constructor(config: AgentConfig) {
    this.name = config.name;
    this.instructions = config.instructions;
    this.tools = config.tools ?? [];
    this.model = config.model ?? "auto";
    this.budget = new Budget(config.budget ?? {});
    this.maxSteps = config.maxSteps ?? DEFAULT_MAX_STEPS;
    this.router = new Router();
    this.apiKey = config.apiKey ?? process.env.OPENAI_API_KEY ?? "";
    this.apiBase = config.apiBase ?? DEFAULT_API_BASE;
    this.telemetry = resolveTelemetry(config.telemetry);
  }

  /**
   * Execute the agent loop with full production harness.
   * Returns an AgentResult with output, cost, steps, and trace URL.
   */
  async run(prompt: string): Promise<AgentResult> {
    const runId = crypto.randomUUID();
    const result: AgentResult = {
      runId,
      output: "",
      steps: [],
      totalCostUsd: 0,
      totalInputTokens: 0,
      totalOutputTokens: 0,
      totalLatencyMs: 0,
      modelUsage: {},
      startedAt: new Date(),
    };

    const messages: ChatMessage[] = [
      { role: "system", content: this.instructions },
      { role: "user", content: prompt },
    ];

    const toolSchemas = this.tools.map(toolToSchema);
    const toolMap = new Map(this.tools.map((t) => [t.name, t]));

    for (let stepNum = 0; stepNum < this.maxSteps; stepNum++) {
      // Gate 1: Route to optimal model
      const currentText = stepNum === 0 ? prompt : lastContent(messages);
      const routing = this.resolveModel(currentText);

      // Gate 2: Check budget
      try {
        this.budget.check(ESTIMATED_COST_PER_CALL);
      } catch {
        result.output = `[Budget exceeded after ${stepNum} steps, spent $${this.budget.spent.toFixed(4)}]`;
        break;
      }

      // Gate 3: Call LLM
      const startTime = performance.now();
      const response = await this.callLLM(routing.model, messages, toolSchemas);
      const latencyMs = Math.round(performance.now() - startTime);

      // Gate 4: Track cost
      const inputTokens = response.usage?.prompt_tokens ?? 0;
      const outputTokens = response.usage?.completion_tokens ?? 0;
      const cost = estimateCost(routing.model, inputTokens, outputTokens);
      this.budget.record(cost);

      const step: Step = {
        id: `${runId}-${stepNum}`,
        model: routing.model,
        tier: routing.tier,
        inputTokens,
        outputTokens,
        costUsd: cost,
        latencyMs,
        cached: false,
        timestamp: new Date(),
      };

      const choice = response.choices?.[0];
      const message = choice?.message;

      // Handle tool calls
      if (message?.tool_calls?.length) {
        messages.push(message as ChatMessage);

        for (const tc of message.tool_calls) {
          const fnName = tc.function.name;
          step.toolCall = fnName;

          let toolOutput: string;
          const tool = toolMap.get(fnName);
          if (tool) {
            try {
              const args = JSON.parse(tc.function.arguments);
              const rawResult = tool.fn(args);
              toolOutput = String(rawResult);
            } catch (err) {
              toolOutput = `[Tool error: ${err}]`;
            }
          } else {
            toolOutput = `[Unknown tool: ${fnName}]`;
          }

          messages.push({
            role: "tool",
            tool_call_id: tc.id,
            content: toolOutput,
          } as ChatMessage);
        }

        recordStep(result, step);
        continue;
      }

      // Final response — no tool calls
      recordStep(result, step);
      result.output = message?.content ?? "";
      break;
    }

    result.finishedAt = new Date();
    result.totalLatencyMs = Math.round(
      result.finishedAt.getTime() - result.startedAt.getTime(),
    );

    // Gate 6: cloud telemetry — fire-and-forget, never blocks the caller.
    // Enabled only when ARCHON_API_KEY is set (or an explicit client was
    // passed). All failures are swallowed inside the client.
    if (this.telemetry?.enabled) {
      this.telemetry.uploadRun(result, this.name);
    }

    return result;
  }

  // ── Private helpers ─────────────────────────────────

  private resolveModel(text: string): RoutingDecision {
    if (this.model !== "auto") {
      return { tier: "standard", model: this.model, reason: "explicit model" };
    }
    return this.router.route(text, this.budget.remaining);
  }

  private async callLLM(
    model: string,
    messages: ChatMessage[],
    tools: object[],
  ): Promise<LLMResponse> {
    const body: Record<string, unknown> = { model, messages };
    if (tools.length > 0) {
      body.tools = tools;
    }

    const response = await fetch(`${this.apiBase}/chat/completions`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`LLM API error ${response.status}: ${errorText}`);
    }

    return (await response.json()) as LLMResponse;
  }
}

// ── Helper types (OpenAI-compatible) ──────────────────

interface ChatMessage {
  role: string;
  content?: string;
  tool_calls?: ToolCall[];
  tool_call_id?: string;
}

interface ToolCall {
  id: string;
  type: "function";
  function: { name: string; arguments: string };
}

interface LLMResponse {
  choices?: Array<{
    message: ChatMessage;
  }>;
  usage?: {
    prompt_tokens: number;
    completion_tokens: number;
  };
}

// ── Pure helper functions ─────────────────────────────

function toolToSchema(tool: ToolDef): object {
  return {
    type: "function",
    function: {
      name: tool.name,
      description: tool.description,
      parameters: {
        type: "object",
        properties: tool.parameters,
        required: Object.keys(tool.parameters),
      },
    },
  };
}

function lastContent(messages: ChatMessage[]): string {
  for (let i = messages.length - 1; i >= 0; i--) {
    if (messages[i].content) return messages[i].content!;
  }
  return "";
}

function recordStep(result: AgentResult, step: Step): void {
  result.steps.push(step);
  result.totalCostUsd += step.costUsd;
  result.totalInputTokens += step.inputTokens;
  result.totalOutputTokens += step.outputTokens;
  result.modelUsage[step.model] = (result.modelUsage[step.model] ?? 0) + 1;
}

/**
 * Normalize the `telemetry` constructor argument.
 *
 *   - `undefined` (default) — create a client that auto-enables when
 *     `ARCHON_API_KEY` is set; otherwise stays disabled.
 *   - `false` — opt out entirely; return `undefined`.
 *   - `true` — same as `undefined` (explicit opt-in to defaults).
 *   - `TelemetryClient` — use the provided instance.
 */
function resolveTelemetry(
  telemetry: TelemetryClient | boolean | undefined,
): TelemetryClient | undefined {
  if (telemetry === false) return undefined;
  if (telemetry instanceof TelemetryClient) return telemetry;
  // `true` or `undefined` — build a default client. It no-ops when the
  // env var is missing, so this is safe for offline users.
  return new TelemetryClient();
}

/** Rough cost estimate based on model name patterns. */
function estimateCost(
  model: string,
  inputTokens: number,
  outputTokens: number,
): number {
  // Pricing per million tokens (input / output)
  const pricing: Record<string, [number, number]> = {
    "gpt-4.1-nano": [0.10, 0.40],
    "gpt-4.1-mini": [0.40, 1.60],
    "gpt-4o": [2.50, 10.0],
    "gpt-4.1": [2.00, 8.00],
    "claude-haiku": [0.80, 4.00],
    "claude-sonnet": [3.00, 15.0],
    "claude-opus": [5.00, 25.0],
    "gemini-flash": [0.15, 0.60],
    "gemini-pro": [1.25, 5.00],
    "o4-mini": [1.10, 4.40],
    "deepseek-r1": [0.55, 2.19],
    "deepseek": [0.27, 1.10],
  };

  // Find the best matching pricing entry
  const lowerModel = model.toLowerCase();
  for (const [pattern, [inputRate, outputRate]] of Object.entries(pricing)) {
    if (lowerModel.includes(pattern)) {
      return (inputTokens * inputRate + outputTokens * outputRate) / 1_000_000;
    }
  }

  // Default: mid-tier pricing
  return (inputTokens * 2.0 + outputTokens * 8.0) / 1_000_000;
}
